alicat/basis.py

```diff
@@ -0,0 +1,346 @@
+"""Python driver for Alicat BASIS mass flow devices, using serial communication.
+
+Distributed under the GNU General Public License v2
+Copyright (C) 2023 NuMat Technologies
+"""
+from __future__ import annotations
+
+import asyncio
+from typing import Any, ClassVar
+
+import nest_asyncio
+
+from .util import Client, SerialClient, _is_float
+
+nest_asyncio.apply()
+class FlowMeter:
+    """Python driver for BASIS Flow Meters.
+
+    [Reference](https://www.alicat.com/wp-content/documents/manuals/DOC-MANUAL-BASIS2.pdf).
+
+    This communicates with the flow meter over a USB or RS-232/RS-485
+    connection using pyserial.
+    """
+
+    # A dictionary that maps port names to a tuple of connection
+    # objects and the refcounts
+    open_ports: ClassVar[dict[int, tuple]] = {}
+    gases: ClassVar[list] = ['Air', 'Ar', 'CO2', 'N2', 'O2', 'N2O', 'H2', 'He', 'CH4']
+
+    def __init__(self, address: str = '/dev/ttyUSB0', unit: str = 'A', baudrate: int = 38400, **kwargs: Any) -> None:
+        """Connect this driver with the appropriate USB / serial port.
+
+        Args:
+            address: The serial port or TCP address:port. Default '/dev/ttyUSB0'.
+            unit: The Alicat-specified unit ID, A-Z. Default 'A'.
+            baudrate: The baud rate of the device. Default 38400.
+        """
+        self.hw: Client = SerialClient(address=address, baudrate=baudrate, **kwargs)
+
+        self.unit = unit
+        self.keys = ['temperature', 'flow', 'totalizer', 'setpoint',
+                     'valve drive', 'gas']
+        self.open = True
+        self.firmware: str | None = None
+
+        ## check if device is basis. if not, raise error
+        async def _is_basis() -> None:
+            self.control_point = await self._check_basis()
+
+        loop = asyncio.get_event_loop()
+        tasks = asyncio.gather(_is_basis())
+        loop.run_until_complete(tasks)

```

Comment:
I'm not sure this is necessary.  It might make sense to put detection of Basis responses in the main driver and throw errors there.

Anyway, if it is necessary it can be done with `asyncio.ensure_future` to get around the constructor not being able to use `await`. I'm pretty sure that will finish before e.g. any reads are made.  If not, it could then be `await`ed by `is_connected` since `await`ing an already-completed task basically does nothing.

Is this making sense, or should I give an example or different explanation?

Reply:
Hi Alex, apologies for the late response. It has been a very busy week... 

That's true (and I assume if anyone is importing the BASIS part of this library then they already have a BASIS device). I'll go ahead and mark that for deletion in the code. Let me know if there's anything else you think could be improved (I'm sure there is!) and I'll put everything into one pull request. Thank you again for being so patient!